### PR TITLE
Geo mechanics branch

### PR DESCRIPTION
**Pull request by Vahid-Galavi**
**Description**
Geomechanics application. This application has been developed to solve geomechanical problems. At the moment, the following features are available:

- Two-phase formulation 
- Dynamic and quasi-static
- K0-procedure to initialise stresses
- User-Defined soil models
- 2D Plane strain and 3D
- Small deformation
- Updated Lagrangian
- Partially saturated
- Reset displacement
- Automatic time stepping

There are several points which need to be improved/solved in the near future. Several issues will be created to address the current short comings of the application, e.g.:
- Instead of having modified copies of structural elements, the structural elements should be derived and then be modified.

### CODE REVIEW

File: applications/GeoMechanicsApplication/custom_conditions/U_Pw_face_load_interface_condition.hpp [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+ 
+    void Initialize(const ProcessInfo& rCurrentProcessInfo) override;
+ 
+//----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
+
+protected:   
+    
+    // Member Variables
+    
+    std::vector<double> mInitialGap;
```
Comment (RiccardoRossi) [inline applications/GeoMechanicsApplication/custom_conditions/U_Pw_face_load_interface_condition.hpp]: just a feedback, but std::vector occupies one double more than Vector

File: applications/GeoMechanicsApplication/custom_conditions/surface_load_3D_diff_order_condition.cpp [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+
+
+// Project includes
+#include "custom_conditions/surface_load_3D_diff_order_condition.hpp"
+
+namespace Kratos
+{
+
+// Default Constructor
+SurfaceLoad3DDiffOrderCondition::SurfaceLoad3DDiffOrderCondition() : GeneralUPwDiffOrderCondition() {}
```
Comment (RiccardoRossi) [inline applications/GeoMechanicsApplication/custom_conditions/surface_load_3D_diff_order_condition.cpp:23]: what is the difference of this one wrt the StructuralMechanicsOne?

File: applications/GeoMechanicsApplication/custom_problemtype/GeoMechanicsApplication.gid/AddConditionsToExcavation.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
+
```
Comment (RiccardoRossi) [inline applications/GeoMechanicsApplication/custom_problemtype/GeoMechanicsApplication.gid/AddConditionsToExcavation.py:1]: I think we agreed that the problemtype should not be committed together with the application, but rather put into a different repository

File: applications/GeoMechanicsApplication/tests/Biaxialshearstresswithlinearelasticmodel.gid/ProjectParameters.json [hunk 4]
```diff
@@ -0,0 +1,180 @@
+{
```
Comment (RiccardoRossi) [inline applications/GeoMechanicsApplication/tests/Biaxialshearstresswithlinearelasticmodel.gid/ProjectParameters.json:1]: I comment here for the file below:

you are including .dll and .so , however this is version dependent and will not owrk depending on where you try to run (just think of trying to run it on a different linux distribution)

File: applications/GeoMechanicsApplication/tests/Oedometer_ULFEM_FIC.gid/ProjectParameters.json [hunk 5]
```diff
@@ -0,0 +1,143 @@
+{
```
Comment (RiccardoRossi) [inline applications/GeoMechanicsApplication/tests/Oedometer_ULFEM_FIC.gid/ProjectParameters.json:1]: just a comment here but which applies to all of the tests: how long does it take for them to run?

do remember that we have a very strict time limit on the CI

**Review by RiccardoRossi — COMMENTED** [review]
in essence my most important question is about the tests.

how long do they take to run? my concern is that they are not very small,

File: applications/GeoMechanicsApplication/custom_conditions/U_Pw_face_load_interface_condition.hpp (hunk 1 above)
Comment (Vahid-Galavi) [reply applications/GeoMechanicsApplication/custom_conditions/U_Pw_face_load_interface_condition.hpp]: Thanks Riccardo! I will replace it with Vector

File: applications/GeoMechanicsApplication/custom_problemtype/GeoMechanicsApplication.gid/AddConditionsToExcavation.py (hunk 3 above)
Comment (Vahid-Galavi) [reply applications/GeoMechanicsApplication/custom_problemtype/GeoMechanicsApplication.gid/AddConditionsToExcavation.py:1]: Thanks Riccardo! I will move it to the other repository!

File: applications/GeoMechanicsApplication/custom_conditions/surface_load_3D_diff_order_condition.cpp (hunk 2 above)
Comment (Vahid-Galavi) [reply applications/GeoMechanicsApplication/custom_conditions/surface_load_3D_diff_order_condition.cpp:23]: I need to check. If these two are the same, then I will remove this one.

File: applications/GeoMechanicsApplication/tests/Biaxialshearstresswithlinearelasticmodel.gid/ProjectParameters.json (hunk 4 above)
Comment (Vahid-Galavi) [reply applications/GeoMechanicsApplication/tests/Biaxialshearstresswithlinearelasticmodel.gid/ProjectParameters.json:1]: This are some tests for user defined material models.

File: applications/GeoMechanicsApplication/tests/Oedometer_ULFEM_FIC.gid/ProjectParameters.json (hunk 5 above)
Comment (Vahid-Galavi) [reply applications/GeoMechanicsApplication/tests/Oedometer_ULFEM_FIC.gid/ProjectParameters.json:1]: All the tests, including nightly tests, take around 165 s (Windows) and 95s (Ubuntu). If the nightly tests are excluded, the time reduces significantly.

**Comment by Vahid-Galavi** [comment]
This PR was created from a wrong branch, in which problemtype exists.

**Pull request closed without merging**